File: nwp-project-back/src/main/java/com/raf/rs/nwp/service/MachineService.java
package com.raf.rs.nwp.service;

import com.raf.rs.nwp.dto.machine.ErrorMessageDTO;
import com.raf.rs.nwp.dto.machine.MachineCreateDTO;
import com.raf.rs.nwp.dto.machine.MachineDTO;
import com.raf.rs.nwp.model.enums.MachineStatus;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.time.LocalDateTime;
import java.time.OffsetDateTime;
import java.util.List;

public interface MachineService {

    Page<MachineDTO> searchMachines(String name, List<MachineStatus> statuses, OffsetDateTime dateFrom, OffsetDateTime dateTo, Pageable pageable);

    void startMachine(Long machineId);

    void stopMachine(Long machineId);

    void restartMachine(Long machineId);

    MachineDTO createMachine(String email, MachineCreateDTO machineDto);

    void destroyMachine(Long machineId);

    void scheduleStartMachine(Long machineId, LocalDateTime scheduledDateTime);

    void scheduleStopMachine(Long machineId, LocalDateTime scheduledDateTime);

    void scheduleRestartMachine(Long machineId, LocalDateTime scheduledDateTime);

    List<String> getAllStatuses();

    Page<ErrorMessageDTO> getAllErrors(String userEmail, Pageable pageable);
}

File: nwp-project-front/src/app/service/machine/machine.service.ts
import {Injectable} from '@angular/core';
import {HttpClient, HttpParams} from '@angular/common/http';
import {Observable} from 'rxjs';
import * as moment from 'moment';
import {ErrorMessageDTO, MachineCreateDTO, MachineDTO, SearchParams} from "../../model/machine";
import {ApiResponse} from "../../model/api-response";
import {Page} from "../../model/page";
import {User} from "../../model/user";
import {API_ENDPOINTS} from "../../constants";

@Injectable({
  providedIn: 'root'
})
export class MachineService {

  constructor(private http: HttpClient) {
  }

  searchMachines(params: SearchParams): Observable<ApiResponse<Page<MachineDTO>>> {
    const transformedParams: SearchParams = {
      ...params,
      dateFrom: params.dateFrom ? moment(params.dateFrom).format() : undefined,
      dateTo: params.dateTo ? moment(params.dateTo).format() : undefined
    };

    return this.http.post<ApiResponse<Page<MachineDTO>>>(
      `${API_ENDPOINTS.MACHINES}/search`,
      transformedParams
    );
  }

  getAvailableStatuses(): Observable<ApiResponse<string[]>> {
    return this.http.get<ApiResponse<string[]>>(
      `${API_ENDPOINTS.MACHINES}/statuses`
    );
  }

  createMachine(machine: MachineCreateDTO): Observable<ApiResponse<User>> {
    return this.http.post<ApiResponse<User>>(
      `${API_ENDPOINTS.MACHINES}/create`,
      machine
    );
  }

  fetchErrors(page: number = 0, size: number = 10): Observable<ApiResponse<Page<ErrorMessageDTO>>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());

    return this.http.get<ApiResponse<Page<ErrorMessageDTO>>>(
      `${API_ENDPOINTS.MACHINES}/errors`,
      {params}
    );
  }

  startMachine(machineId: number): Observable<ApiResponse<string>> {
    return this.http.put<ApiResponse<string>>(
      `${API_ENDPOINTS.MACHINES}/${machineId}/start`,
      null
    );
  }

  restartMachine(machineId: number): Observable<ApiResponse<string>> {
    return this.http.put<ApiResponse<string>>(
      `${API_ENDPOINTS.MACHINES}/${machineId}/restart`,
      null
    );
  }

  stopMachine(machineId: number): Observable<ApiResponse<string>> {
    return this.http.put<ApiResponse<string>>(
      `${API_ENDPOINTS.MACHINES}/${machineId}/stop`,
      null
    );
  }


  destroyMachine(machineId: number): Observable<ApiResponse<string>> {
    return this.http.put<ApiResponse<string>>(
      `${API_ENDPOINTS.MACHINES}/${machineId}/destroy`,
      null
    );
  }
}
